examples/direct_fidelity_estimation.py

```diff
@@ -52,54 +58,108 @@ def compute_characteristic_function(circuit: cirq.Circuit,
 
+def sample_pauli_states(n_qubits: int, n_trials: int, physical: bool):
+    pauli_gates = [cirq.I, cirq.X, cirq.Y, cirq.Z]
+    if physical:
+        return [[pauli_gates[j]
+                 for j in np.random.choice(4, n_qubits)]
+                for i in range(n_trials)]
+    else:
+        return itertools.product(pauli_gates, repeat=n_qubits)
+
+
 def direct_fidelity_estimation(circuit: cirq.Circuit, qubits: List[cirq.Qid],
-                               noise: cirq.NoiseModel, n_trials: int):
+                               noise: cirq.NoiseModel, n_trials: int,
+                               physical: bool):
     # n_trials is upper-case N in https://arxiv.org/pdf/1104.3835.pdf
 
     # Number of qubits, lower-case n in https://arxiv.org/pdf/1104.3835.pdf
     n_qubits = len(qubits)
+    d = 2**n_qubits
 
     # Computes for every \hat{P_i} of https://arxiv.org/pdf/1104.3835.pdf,
     # estimate rho_i and Pr(i). We then collect tuples (rho_i, Pr(i), \hat{Pi})
     # inside the variable 'pauli_traces'.
     pauli_traces = []
-    for P_i in itertools.product([cirq.I, cirq.X, cirq.Y, cirq.Z],
-                                 repeat=n_qubits):
-        rho_i, Pr_i = compute_characteristic_function(circuit,
-                                                      P_i,
-                                                      qubits,
-                                                      noise=None)
+
+    simulator = cirq.DensityMatrixSimulator()
+    if physical:
+        density_matrix = None
+    else:
+        # rho in https://arxiv.org/pdf/1104.3835.pdf
+        density_matrix = cast(cirq.DensityMatrixTrialResult,
+                              simulator.simulate(circuit)).final_density_matrix
+
+    for P_i in sample_pauli_states(n_qubits, n_trials, physical):
+        rho_i, Pr_i = compute_characteristic_function(circuit, P_i, qubits,
+                                                      physical, simulator,
+                                                      density_matrix)
         pauli_traces.append({'P_i': P_i, 'rho_i': rho_i, 'Pr_i': Pr_i})
 
-    assert len(pauli_traces) == 4**n_qubits
+    if physical:
+        assert len(pauli_traces) == n_trials
+    else:
+        assert len(pauli_traces) == 4**n_qubits
 
     p = [x['Pr_i'] for x in pauli_traces]
-    assert np.isclose(sum(p), 1.0, atol=1e-6)
+    if physical:
+        # Sometimes, the probabilities add up to 0.0 just because of randomness
+        # and so we hard-code a 0 fidelity in this case.
+        if sum(p) == 0.0:
+            return 0.0
+    else:
+        assert np.isclose(sum(p), 1.0, atol=1e-6)
 
     # The package np.random.choice() is quite sensitive to probabilities not
     # summing up to 1.0. Even an absolute difference below 1e-6 (as checked just
     # above) does bother it, so we re-normalize the probs.
     inv_sum_p = 1 / sum(p)
     norm_p = [x * inv_sum_p for x in p]
 
+    simulator = cirq.DensityMatrixSimulator(noise=noise)
+    if physical:
```

Comment:
You say `if physical` three times in this method. This suggests you should have two methods. The large common parts of this method can be extracted into helper methods. You also say `if physical` in other places... if this were not an example, I would recommend extracting a class with methods corresponding to each of these `if` blocks.

Reply:
With Matthew's comment, I think it's only in one place now. Would you like to have the section of the code under the single 'if sampled:' in its separate function. I could have a shared 'for' loop.

Not feeling strongly about it.